### ISSUE TITLE
Perpetually loading landing page of metaplex storefront

### ISSUE THREAD

**Issue by soulsayer9029** [metaplex-foundation/metaplex#1339]
**The Problem**
Metaplex web application(storefront) is not working and as the landing page loads in perpetuity

**To Reproduce**
1. Clone the metaplex repository 
2. cd to metaplex/js
3. yarn install
4. yarn bootstrap
5. yarn start
6. Go to localhost:3000 and see the output
 
The loading sign is just not going away and the page loads perpetually

**Expected behavior**
There should not be a perpetual loading sign 

**Screenshots**
![Screenshot from 2021-12-28 18-06-18](https://user-images.githubusercontent.com/60864910/147567237-5400f32e-9e04-43a1-8d24-4729e646dd65.png)

**Desktop:**
 - OS: Ubuntu 18
 - Browser Mozilla Firefox

**Comment by ChristianFillon** [comment]
Hi
Same thing with Chrome, using latest branch release/v1.1.0.
Thank's for this description, I thought it was a mistake on my part. I also found a description here: https://issueexplorer.com/issue/metaplex-foundation/metaplex/512.
Christian

**Comment by neurofoo** [comment]
@soulsayer9029 did you already init the store? if so, are you on devnet? Sometimes I've had to force the network by appending `?network=devent` in the URL.

**Comment by ChristianFillon** [comment]
Hi. 
Happy new year.
I tried this morning from home, with a good network. After a long time, I had an "out of memory" message on the Chrome tab, but not all the memory is not used on the computer. It seems it is a Chrome limitation.
Using this URL: http://localhost:3000/#/?network=devnet

**Comment by neurofoo** [comment]
hmm.

A few things. You might have already done these. (caveats, I'm new to this codebase too!)

Is your .env empty? Make sure you don't have anything defined for `REACT_APP_STORE_OWNER_ADDRESS_ADDRESS` 

For the devtools console, you should see output about store owner and address. Those should be undefined. If they aren't, then I think that the code thinks the store has been initialized. 

Try in incognito mode. I just tried and even with a store configured in incognito mode I get the init screens to show. Not sure why it would be doing that in incognito, but this might be a base test too.

**Comment by aheff** [comment]
Has anyone made more progress with this? I am also new to this whole Metaplex thing. I managed to initialise a store on the mainnet beta and even create a few NFTs. The loading phase was taking a while; now it won't load at all and get's stuck on the 'just loading' screen like the people above.

Running on Mac OS.

**Comment by ChristianFillon** [comment]
@neurofoo
I checked your idea of leaving the .env file empty. In this situation, there is no longer the infinite loading because there is no attempt to load because of the incomplete configuration. But then it is not possible to connect the wallet because there must be a check between the wallet address and the address given in the .env file. This test shows that the error is more precisely in the step of loading information to the Solana network, the "devnet" in my case.
All the tutorials specify that the wallet address must be entered in the .env file, I think this instruction must be respected and that the problem is elsewhere.

**Comment by ChristianFillon** [comment]
Any news on this subject

**Comment by stegaBOB** [comment]
We are aware of the issue and are working on a fix

**Comment by DeltaMichael** [comment]
Still facing this, even tried deploying an older commit (tag v1.0.0), but still running into the issue. https://deltamichael.github.io/metaplex#/

**Comment by MrMizz** [comment]
minting crashes my chrome browser. general usage is horribly slow as well. I guess back to custom development . . . 
```
breadcrumbs.ts:111 Warning: React has detected a change in the order of Hooks called by LaunchStep. This will lead to bugs and errors if not fixed. For more information, read the Rules of Hooks: https://reactjs.org/link/rules-of-hooks

   Previous render            Next render
   ------------------------------------------------------
1. useState                   useState
2. useState                   useState
3. useEffect                  useEffect
4. useEffect                  useEffect
5. undefined                  useContext
   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

    at LaunchStep (webpack-internal:///./src/views/artCreate/index.tsx:1781:54)
    at div
    at eval (webpack-internal:///../../node_modules/antd/es/grid/col.js:49:62)
    at div
    at eval (webpack-internal:///../../node_modules/antd/es/grid/row.js:45:34)
    at ArtCreateView (webpack-internal:///./src/views/artCreate/index.tsx:72:83)
    at component
    at Route (webpack-internal:///../../node_modules/react-router/esm/react-router.js:450:29)
    at Switch (webpack-internal:///../../node_modules/react-router/esm/react-router.js:652:29)
    at main
    at Basic (webpack-internal:///../../node_modules/antd/es/layout/layout.js:72:25)
    at Adapter (webpack-internal:///../../node_modules/antd/es/layout/layout.js:55:66)
    at section
    at BasicLayout (webpack-internal:///../../node_modules/antd/es/layout/layout.js:87:63)
    at Adapter (webpack-internal:///../../node_modules/antd/es/layout/layout.js:55:66)
    at section
    at BasicLayout (webpack-internal:///../../node_modules/antd/es/layout/layout.js:87:63)
    at Adapter (webpack-internal:///../../node_modules/antd/es/layout/layout.js:55:66)
    at _c (webpack-internal:///./src/components/Layout/index.tsx:68:27)
    at ConfettiProvider (webpack-internal:///./src/components/Confetti/index.tsx:24:3)
    at LoaderProvider (webpack-internal:///./src/components/Loader/index.tsx:21:3)
    at MetaProvider (webpack-internal:///../common/dist/lib/contexts/meta/meta.js:41:25)
    at StoreProvider (webpack-internal:///../common/dist/lib/contexts/store.js:27:26)
    at CoingeckoProvider (webpack-internal:///./src/contexts/coingecko.tsx:50:3)
    at SPLTokenListProvider (webpack-internal:///./src/contexts/tokenList.tsx:36:3)
    at AccountsProvider (webpack-internal:///../common/dist/lib/contexts/accounts/accounts.js:116:29)
    at WalletModalProvider (webpack-internal:///../common/dist/lib/contexts/wallet.js:83:32)
    at WalletProvider (webpack-internal:///../../node_modules/@solana/wallet-adapter-react/lib/WalletProvider.js:25:27)
    at WalletProvider (webpack-internal:///../common/dist/lib/contexts/wallet.js:116:27)
    at ConnectionProvider (webpack-internal:///../common/dist/lib/contexts/connection.js:57:31)
    at Providers (webpack-internal:///./src/providers.tsx:28:3)
    at Router (webpack-internal:///../../node_modules/react-router/esm/react-router.js:81:30)
    at HashRouter (webpack-internal:///../../node_modules/react-router-dom/esm/react-router-dom.js:103:35)
    at Routes
    at App
    at LoadableImpl (webpack-internal:///../../node_modules/next/dist/shared/lib/loadable.js:107:36)
    at App
    at div
    at App (webpack-internal:///./src/pages/_app.tsx:25:3)
    at ErrorBoundary (webpack-internal:///../../node_modules/@next/react-dev-overlay/lib/internal/ErrorBoundary.js:26:47)
    at ReactDevOverlay (webpack-internal:///../../node_modules/@next/react-dev-overlay/lib/internal/ReactDevOverlay.js:86:23)
    at Container (webpack-internal:///../../node_modules/next/dist/client/index.js:225:1)
    at AppContainer (webpack-internal:///../../node_modules/next/dist/client/index.js:604:3)
    at Root (webpack-internal:///../../node_modules/next/dist/client/index.js:734:3)
    ```

**Comment by shudhanshu-in** [comment]
Any Update on this issue ???

**Comment by mer0x** [comment]
any news ?

**Comment by arshia21** [comment]
For me the problem was gone when I changed to mainnet and added some actual SOL to the market owner address, and also to the Phantom wallet address

**Comment by v-raja** [comment]
Did you have to pay actual SOL to initialize the store or did just having
some SOL in the wallet make it work?

**Comment by mer0x** [comment]
Are you sure it s not just cached on local? For me when I try to access the website from a new browser it takes 1-2 minutes for first connection, 10-15 seconds after this, It seems this problem doesn t have any fix yet

**Comment by arshia21** [comment]
Yes, I had to pay around 0.1 actual SOL to initialise the store. Still it takes 1-2 minutes for first connection, but it i couldn't get it connected on devnet at all

**Comment by mapleflavor** [comment]
I just ran a build and same issues as you are all having. I tested with SOL in the wallet and still having issues. I'm running the test on github pages with no luck and perpetual loading still.

**Comment by stegaBOB** [comment]
Sorry for the lack of a quicker update. The issue right now is the speed at which you can query NFTs off chain. Sorting and querying all the NFTs from the blockchain isn't a really fast operation at this time since there's ~6,000,000 Solana NFTs it needs to sort through. Getting an indexer of some kind up and running is a solution that is currently been worked on which would solve this issue. I know this isn't the best of news, but unfortunately there isn't really a better option until a faster indexer is ready. Until then the initial loading time is just going to be slower than we'd like.

**Comment by ChristianFillon** [comment]
So metaplex is dead ? No solution to use it ?

**Comment by stegaBOB** [comment]
For the time being with storefront V1 (we are hard at work on a fully revamped v2), loading times are going to be slower (10s+ on fast rpc endpoints for larger collections) than we would like. There are quite a lot more things that Metaplex does than storefronts though, so I wouldn't say Metaplex is dead in the slightest. Regardless, the storefront will be getting better soon.

**Comment by stegaBOB** [comment]
Unlike on Ethereum or other EVM chains, all NFTs on Solana are native SPL tokens. This means that in order to get the NFTs you want, the RPC nodes have to look through every Token Metadata account to find the NFTs with the matching metadata that you are looking for. This is why indexing by specific byte offsets will be really important in the future.

**Comment by v-raja** [comment]
Apologies if this is a naive question but how will the v2 storefront solve this?

**Comment by stegaBOB** [comment]
In theory, with more effecient react, a faster UI, and ideally an indexer that works will be reased by then.

**Comment by MrMizz** [comment]
@stegaBOB however on the solana chain you can easily query for all SPL tokens associated/minted by a given public-key.

**Comment by conman0552** [comment]
I do not think V2 has solved this issue. 

I am running v2 (or should be) on a github pages setup and experience quite slow landing on the main page. Typically it takes 2-3 minutes to load in an incognito window on Chrome. If I am in a regular chrome instance, it may not load at all (note I close the phantom wallet prompt and the "error" prompt shows up quickly, so I know that response time should be semi-fast.

I'm wondering if this is a git pages issue or another altogether. I'll throw the same instance together into a webserver and report back.

**Comment by nullmicgo** [comment]
After waited for a while, it shows the following error
Error: 503 Server Overloaded: {"jsonrpc":"2.0","error":{"code":503,"message":"Server Overloaded"},"id": null}

 

![image](https://user-images.githubusercontent.com/1070485/154798478-d327e293-4ec7-4562-a064-f9f55aab7bc5.png)

**Comment by facetimer** [comment]
Any news?

**Comment by andrereus** [comment]
so this being closed means there isn't an actual solution for this currently and the storefront is therefore useless?

**Comment by andrereus** [comment]
how are some popular stores who are using metplex solving this problem? they seem to be faster

**Comment by stegaBOB** [comment]
> I do not think V2 has solved this issue. 
> 
> I am running v2 (or should be) on a github pages setup and experience quite slow landing on the main page. Typically it takes 2-3 minutes to load in an incognito window on Chrome. If I am in a regular chrome instance, it may not load at all (note I close the phantom wallet prompt and the "error" prompt shows up quickly, so I know that response time should be semi-fast.
> 
> I'm wondering if this is a git pages issue or another altogether. I'll throw the same instance together into a webserver and report back.

V2 isn't out yet. That's probably why it hasn't solved the issue haha. As of now the only workaround that doesn't involve a backend would be to build a custom caching/indexing layer into your storefront. This is what Holaplex has done I believe. Without that indexing layer, the problem is the slow read speed for RPCs. Until a caching/indexing layer is built on top of the rpcs, the speed of the current storefront will unfortunately just continue to slow down as the number of mints increases. A potential improvement in the meantime would be to replace all the current getProgramAccounts with a GPA that has a data size of 0. This returns the request a bit faster from there you can use getMultipleAccounts with up to 100 at a time to get the full necessary data (aka paginate it!).  

Orrrrr if you want to do something a bit different and really dig into the weeds you can hardcode the mint addresses into the GPA calls and update those when you want to either mint a new token or something like that (or maybe use the mint list when a user goes to your site and then have the GPA calls run in the background  and update your mint list in a backend database which can then update the mintlist on the frontend if needed). 

All of these are short term "solutions" that are out of the scope of what we will do in the short term, but we would gladly review/merge any PR that implements a solution like that.

**Comment by andrereus** [comment]
this is an interesting post I found regarding the RPCs https://moralis.io/ethereum-rpc-nodes-what-they-are-and-why-you-shouldnt-use-them/

**Comment by EzanAsif** [comment]
Still getting this issue. Any solution for this?

**Comment by flatplate** [comment]
I had the same issue and the problem for me was that I had the wrong node version. Readme says version v14.17.6 and I had v16. After installing the proper version using nvm the problem was fixed.

**Comment by thatotherkid86** [comment]
Is there a fix for this? On localhost it takes 2-3 minutes and on Vercel, even longer.

**Comment by miukki** [comment]
i use 14.18.1 version and I have perpetual loading. and i believe it is related fetching nft is very long process

**Comment by miukki** [comment]
why you closed issue? @stegaBOB

**Comment by deeplearn-art** [comment]
Since this issue makes storefront unusable, it would be courteous to reference it in the README until fixed, so as not to disappoint people who wish to open up a custom NFT store on mainnet.

**Comment by ShaneDeconinck** [comment]
I'm also not able to get the Metaplex storefront running

**Comment by mnedelko** [comment]
Hello. Was wondering whether there is an update on this issue. I also experience the perpetually loading screen.

**Comment by ljuliandicarlo** [comment]
I've tried a few things: deleting .env / .env.production, clearing cash on both browser and Phantom, and reinstalling dependencies. Nothing worked -- still experiencing excessive load times at home page. Anyone know what the issue is here?

**Comment by mikecann** [comment]
Sooo.. why is this issue closed? It definately is not solved using the latest version..

![image](https://user-images.githubusercontent.com/215033/162356959-71dd43a3-863f-43c7-8e11-c588802a011d.png)

**Comment by jester91** [comment]
Hi! I'm having the same issue with v2. I tried to deploy the code to GitHub pages, and run on localhost, and localhost/devnet but still facing the same issue. 
![image](https://user-images.githubusercontent.com/50679897/164913282-d8ba0a4b-465b-47fb-b724-fb4133a77f93.png)
Tried to run the page(yarn start) from metaplex/js folder and metaplex/js/packages/web

**Comment by planktonrobo** [comment]
Same issue months later. Should not be closed

**Comment by Nedi11** [comment]
I also have the same issue.

**Comment by Nedi11** [comment]
> I also have the same issue.

I found this weird temporary fix to get the store initialized.

I inspect element removed the loading div and started wandering around the site (opened different pages: my items, collections, creators) and when I got back to the main page it loaded correctly, with the init button like in the image shown in docs. Initialized the store, added the wallet to .env and now it works

**Comment by gabazureus** [comment]
Any news? Can't use it effectively :(
Someone can develop a solution to this loading please, sir

**Comment by Marchccc** [comment]
so disappointed

**Comment by asifhassansoc** [comment]
any solution ?

**Comment by tastycrackers** [comment]
same here..very frustrating. Devs, can't solve this ? any improvement will be much appreciated!

**Comment by coladimeji** [comment]
My metaplex was slow when  deployed in localhost and to preview- production 
It was suggested I use API, and I did, the local host is loading in less than 5 seconds but the preview-prod  loads  on  the  average 2.48 minutes
any solution/ suggestion

**Comment by tastycrackers** [comment]
Found this:
https://stackoverflow.com/questions/72300280/metaplex-storefront-on-vercel-perpetual-loading-problem

**Comment by maksimonishchenko** [comment]
lol only 1 needed metaplex solution atm not working at all from initial base not 1 version not second only 1 enthuisiast keep things rolling so a test could be succesfull lol nswered May 19 at 7:53
user avatar
[neft.world](https://stackoverflow.com/users/19120055/neft-world)

**Comment by coladimeji** [comment]
The perpetual loading/ performance issue is over.
**Solution**
Check initial installations
make sure you have Rust, Node, Docker, Solana CLI, yarn
get API and configured it.
NOTE: It is important to have RUST when dealing with metaplex, this is one of the main issues I learned.
get or clone metaplex git you want to implement
Go to packages/web/.env
if you are deploying to devnet and similarly,
go to packages/web/.env.production
if you are deploying to production, production preview, or production test environments
Inside any of the folders, you intend to deploy
In my case its prod preview
Insert your Solana Master/Owners Address 
Insert your Solana RPC network URL
in my case I used API, 
NOTE: If you are not using API use the Solana RPC address  that is in the folder by default
Save entries
Navigate back to /js/packages/package.json
on line 59 - change metaplex to your github username and project name
on line 60 - remove any sentences before : "yarn build && yarn deploy:gh",
Save changes
Navigate back to 
/js
insert yarn install **yarn bootstrap ** yarn build
To deploy
navigate to packages/web
insert yarn deploy
NOTE: make sure if you are using dev, you must have got your API to point to dev, if not it will be using your prod. Solana coin to process tests.
